### ISSUE TITLE
Proposal `-<<` (`distribute`)

### ISSUE THREAD

**Issue by emacksnotes** [magnars/dash.el#407]
Proposal `-<<` (`distribute`)

`-<<` looks like a fork in the pipeline, with the same `arg` getting fed in to multiple `forms`.

```
(defmacro -<< (x forms)
  `(list ,@(->> forms
                (--map `(->> ,x ,it)))))
```

With the above definition, the form

```
(-<< '(H S L)
     ((--map (intern (format "INDEX-OF-%s-IN-HSLUV" it)))
      (funcall (-compose (-cut -iota <> 0) #'length))))
```

expands to 

```
(list
 (->> '(H S L) (--map (intern (format "INDEX-OF-%s-IN-HSLUV" it))))
 (->> '(H S L) (funcall (-compose (-cut -iota <> 0) #'length))))

```

and evaluates to 

```
((INDEX-OF-H-IN-HSLUV INDEX-OF-S-IN-HSLUV INDEX-OF-L-IN-HSLUV)
 (0 1 2))
```

Here is a non-trivial example,

```
(->> (-<< '(H S L)
          ((--map (intern (format "INDEX-OF-%s-IN-HSLUV" it)))
           (funcall (-compose (-cut -iota <> 0) #'length))))
     (apply #'-zip-lists))
```

evaluates to

```
((INDEX-OF-H-IN-HSLUV 0) (INDEX-OF-S-IN-HSLUV 1)
 (INDEX-OF-L-IN-HSLUV 2))
```

Two `<<` in `-<<` signifies that two `>>`-s are used in it macro expansion.   So, `-<<` distribute at the last position.   

Likewise, a `-<` could be provided, which uses `->` in its macro expansion, and distributes its argument to the first position.

**Comment by emacksnotes** [comment]
Few more examples

```
(defun -avg (values)
  (->> (-<< values
	    ((apply #'+)
	     (funcall (-compose #'float #'length))))
       (apply #'/)))
```

```
(-avg (-iota 10))
```

```
4.5
```

```
(defun -min-avg-and-max (values)
  (-<< values
       (-min -avg -max)))
```

```
(-min-avg-and-max (-iota 10))
```

```
(0 4.5 9)
```

**Comment by emacksnotes** [comment]
If there is an existing elegant way of doing `-<<`, I will be happy to hear about it, and use it.

**Comment by emacksnotes** [comment]
May be it shouldn't be a macro

```
(defmacro -<< (x forms)
  `(list ,@(->> forms
                (--map `(->> ,x ,it)))))
``

just a defun like so

```
(defun -<< (x fs)
  (->> fs
       (--map (funcall it x))))
```

You can close this bug if you want (or leave it open), I may keep adding note as I refine my thinking.  For now, this `distribute` is just an "intuition";  I wanted to get the idea out of my head ... so that I am not stuck.

The idea of `distribute` is it transforms a `value`, in to `a list of values`

**Comment by kjambunathan** [comment]
There is a similar request to this one at [-<< / swiss-arrows?](https://github.com/magnars/dash.el/issues/228).

But my request is independent of that note ... Since I don't use Haskell, or Clojure or Common Lisp even my comments are not influenced by what passes for norm in those languages.

I think it should be a `defmacro`.  And it can be used for `iterations`.  `-<<` as intended here, builds multiple levels of tree.  Threading macros `->>` add siblings to the tree.

`-<<` as intended could be used for iterations, on multiple dimensions.  The nodes one level deep, or iteration on first dimension,  The nodes that are two level deep, iterate on the other dimension.

For a concrete use case, think of generating colors in HSLUV space.  First level of `-<<` will generate a single dimensional vector of all possible values of `H`. The second level of `-<<` will generate a two dimensional vector where the `H` value is fixed by its parent node, and `S` values spans the entire spectrum.

In other words, when you nest `-<<`, the values that are generated exponentially blow up.

Leave the bug open, I am using it a scratchpad to clarify my own thinking or as an invitation to brainstorm.  I have some familiarity of working with `R` language--data frame, tidyverse and stuff---and I "vaguely" recollect there being a counterpart to `-<<` in `R`.  Since I don't use `R` for day to day use, I have to consult the docs or dig in to my notes to clarify my own request.

I will leave this "vague remark" to set some context for the request, so that others can dive in with their own ideas.
